src/filebrowser/dialogs.ts

```diff
@@ -172,6 +217,145 @@ class OpenWithHandler extends Widget {
 
 
 /**
+ * A widget used to create a file using a creator.
+ */
+class CreateFromHandler extends Widget {
+  /**
+   * Construct a new "create from" dialog.
+   */
+  constructor(model: FileBrowserModel, manager: DocumentManager, creatorName: string) {
+    super({ node: Private.createCreateFromNode() });
+    this.addClass(FILE_DIALOG_CLASS);
+    this._model = model;
+    this._manager = manager;
+    this._creatorName = creatorName;
+
+    // Check for name conflicts when the input changes.
+    this.input.addEventListener('input', () => {
+      let value = this.input.value;
+      if (value !== this._orig) {
+        for (let item of this._model.items) {
+          if (item.name === value) {
+            this.addClass(FILE_CONFLICT_CLASS);
+            return;
+          }
+        }
+      }
+      this.removeClass(FILE_CONFLICT_CLASS);
+    });
+  }
+
+  /**
+   * Dispose of the resources used by the widget.
+   */
+  dispose(): void {
+    this._model = null;
+    this._manager = null;
+    super.dispose();
+  }
+
+  /**
+   * Get the input text node.
+   */
+  get input(): HTMLInputElement {
+    return this.node.getElementsByTagName('input')[0] as HTMLInputElement;
+  }
+
+  /**
+   * Get the kernel dropdown node.
+   */
+  get kernelDropdown(): HTMLSelectElement {
+    return this.node.getElementsByTagName('select')[0] as HTMLSelectElement;
+  }
+
+  /**
+   * Show the createNew dialog.
+   */
+  show(): Promise<Widget> {
+    return showDialog({
+      title: `Create New ${this._creatorName}`,
+      body: this.node,
+      okText: 'CREATE'
+    }).then(result => {
```

Comment:
This promise can exit without returning a widget. It should probably explicitly return `null` instead of just falling off. Audit return conditions.